### ISSUE TITLE
Forms inserted multiple times in db with match exactly

### ISSUE THREAD

**Issue by lognaturel** [getodk/collect#5147]
#### Software and hardware versions 
Collect v2022.2.2

#### Problem description
We've sometimes had reports of users getting duplicate formid/version errors. We thought the issue was mostly related to the server so for v2022.2.2, @grzesiek2010 made sure that on download a formid/version match results in an update, not a duplication. 

I just got the error with v2022.2.2 and it now looks like some kind of race condition to me.

#### Steps to reproduce the problem
I can't reproduce but here's what I did:
- Fresh 2022.2.2 install with no projects configured
- Scanned in the QR code for Foo at https://test.getodk.cloud/#/projects/149/app-users (access to one small form, one bigger one)
- Opened Filled Blank Form (I don't remember doing this particularly quickly. I've tried tapping on it as quickly as I can and can't reproduce)
- Filled out the `username` form, on finalization, got the error
- Switched form update mode to "manual", looked at Delete Saved Form, saw each form twice
- Looked at forms.db and saw
  ```
  1|2021 WHO Verbal Autopsy instrument||va_who_2021|2021102801|b8cb32fd35df50b9a0fbe86b5df5706d|1653063781657|2021 WHO Verbal Autopsy instrument-media|2021 WHO Verbal Autopsy instrument.xml||||b8cb32fd35df50b9a0fbe86b5df5706d.formdef||||
  2|2021 WHO Verbal Autopsy instrument||va_who_2021|2021102801|b8cb32fd35df50b9a0fbe86b5df5706d|1653063781657|2021 WHO Verbal Autopsy instrument-media|2021 WHO Verbal Autopsy instrument.xml||||b8cb32fd35df50b9a0fbe86b5df5706d.formdef||||
  3|username||username||b4b827d9630d452a12a0e78434457807|1653063781748|username-media|username.xml||||b4b827d9630d452a12a0e78434457807.formdef||||
  4|username||username||b4b827d9630d452a12a0e78434457807|1653063781748|username-media|username.xml||||b4b827d9630d452a12a0e78434457807.formdef||||
  ```

That App User code had definitely never been scanned by that install.

#### Expected behavior
This should never be a possible state.

#### Other information 
It feels like there's some sort of race condition we should identify and fix. It also feels like there should be a uniqueness constraint at the db level on md5 hash.

**Comment by lognaturel** [comment]
It just happened again and I got 4 duplicates of the one form in that project. I don't know what gave me this idea because again I can't reproduce but I started wondering whether it could have to do with the QR code being scanned more than once (not being debounced).

It’s very interesting that I’d never experienced this before and ran into it twice today. I’m in a hurry and going through several demo forms so that points to something speed-related. It’s just very weird that I can’t reproduce when I try to go really fast.

**Comment by grzesiek2010** [comment]
Hmm I can try to reproduce.

**Comment by grzesiek2010** [comment]
I tried to reproduce the problem and I was able to do that at least once. Interestingly I ended up with 2 forms inserted even **three times**: 
![image](https://user-images.githubusercontent.com/3276264/170230609-78455ceb-1444-414c-b828-169f71e82241.png)

I'm going to spend some time on it maybe I will be lucky and able to track it down. If not I will just make md5 column unique (as we discussed yesterday) so that it crashes the app if things like that happen.

**Comment by grzesiek2010** [comment]
Ok I got it.
The problem is that when a user scans a QR code settings are updated one by one. [This starts tasks like `match_exactly` multiple times](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/application/CollectSettingsChangeHandler.java#L31) (4-5 from what I've noticed).
Of course before starting a new task we cancel the existing one but if it does something it [can't be canceled immediately](https://developer.android.com/reference/androidx/work/WorkManager):
>  Cancels all unfinished work in the work chain with the given name. Note that cancellation is a best-effort policy and work that is already executing may continue to run. 

generally we are protected because we use [formsLock.withLock](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/formmanagement/FormsUpdater.kt#L79) but it looks like if two tasks try to acquire it at the same time or when the difference in time is very small it doesn't protect us and lets both tasks to do the job.

As a result multiple tasks download and save forms at the same time.

That was a tricky one... Now as we know what's happening I will try to find the best solution to handle it.

**Comment by lognaturel** [comment]
Thank you, @grzesiek2010!! This has been a long-standing mystery.

**Comment by lognaturel** [comment]
Is your sense that it has to do with something like the angle at which you initiate code scanning and how quickly you move the device away? Like I said, I suspected the QR code scanning might be the issue but even with that idea in mind I couldn’t repro deliberately. 

Is there any way to debounce the QR code scanning? It’s also very strange when a code has invalid settings. In that case, a toast is shown so the same view keeps being shown and if you have the camera beep on it beeps a bunch of times. 

I vaguely remember we’d looked at this and determined it wasn’t possible without slowing down the common case. If that’s true, it’s not worth it.

**Comment by grzesiek2010** [comment]
> Is your sense that it has to do with something like the angle at which you initiate code scanning and how quickly you move the device away? Like I said, I suspected the QR code scanning might be the issue but even with that idea in mind I couldn’t repro deliberately.

No it has nothing to do with the issue. The code is scanned once but it contains multiple settings. Then [scheduleUpdates](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/application/CollectSettingsChangeHandler.java#L31) is called a few times (4-5) for a single QR code (becasue we do that for every single new setting) what starts multiple `SyncFormsTaskSpec`. Then depends on timing our protections might not work well and not block duplicate tasks.

The only thing that might increase your chance of reproduction is the number of forms I guess. I have a project with 7 forms and today I was able to reproduce the issue in 1 of 5 tries let's say but I've never had more than 2 duplicate forms at the same time. So during downloading 7 forms only up to 2 were duplicate what indicates that the more forms you have the bigger chance of reproduction is, but that is just a guess.

> Is there any way to debounce the QR code scanning? 

in this case as I said it has noshing to do with the issue because the code is scanned once.

**Comment by lognaturel** [comment]
Thanks for the thorough explanation, I'm with you now.

Just in case you haven't already explored this, seems to me that these update tasks should use [`enqueueUniqueWork`](https://developer.android.com/reference/androidx/work/WorkManager#enqueueUniqueWork(java.lang.String,%20androidx.work.ExistingWorkPolicy,%20java.util.List%3Candroidx.work.OneTimeWorkRequest%3E)) or some other strategy to guarantee uniqueness.

**Comment by grzesiek2010** [comment]
> Just in case you haven't already explored this, seems to me that these update tasks should use [enqueueUniqueWork](https://developer.android.com/reference/androidx/work/WorkManager#enqueueUniqueWork(java.lang.String,%20androidx.work.ExistingWorkPolicy,%20java.util.List%3Candroidx.work.OneTimeWorkRequest%3E)) or some other strategy to guarantee uniqueness.

We already use this. In our case it's `enqueueUniquePeriodicWork`. It seems like it doesn't work well. I guess like in case of [cancelUniqueWork](https://developer.android.com/reference/androidx/work/WorkManager#cancelUniqueWork(java.lang.String)) work that is already executing may continue to run.
